[QT] Bugfix: ensure tray icon menu is not empty

jamoes opened the pull request:
I am experiencing a bug in which the tray icon menu does not contain any content. This change resolves this bug by moving the `trayIcon->show()` call to be made after the `trayIcon->setContextMenu(trayIconMenu)` call.

krisives commented:
This bug effects me as well.

jamoes commented:
It looks like all the Travis CI builds and checks succeeded except for one, which failed with this message:

> No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.

Perhaps just re-running it will resolve it? I don't see any reason why this change would cause the build to timeout in that manner.

sickpig commented:
@jamoes would you mind share the travis URL where the error is reported? 

lately we had an error like that but wasn't unrelated to `bitcoin-qt`, it was cause by a new c++ unit test we recently added and that increase the time needed to perform the `make check` command on windows machine. 

Such issue has been fixed on `dev` branch a few days ago. 

Today @gandrewstone merged the same fix for the `release` branch. 

To verify is this is the actual problem I think that you should just rebase your branch and push it again. This would automatically update the PR and trig a new travis test.

jamoes commented:
Thnaks @sickpig! The original failing Travic CI build results can be seen here: https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/builds/197878900

I've just rebased with the `release` branch, so hopefully that fixes things.

krisives commented:
Other than compiling from source how long will it take until users have this fix available to them in binary form? I don't know your guys release cycle or if any PPAs or similar are setup for automatic builds.

sickpig commented:
once we have the code committed I'm going to update Ubuntu ppa nightly
channel.

To have it on a stable release you must wait until 1.0.1 or 1.0.0.2 is out
I suppose.


On Mon, Feb 6, 2017 at 7:24 PM, Kristopher Ives <notifications@github.com>
wrote:

> Other than compiling from source how long will it take until users have
> this fix available to them in binary form? I don't know your guys release
> cycle or if any PPAs or similar are setup for automatic builds.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/271#issuecomment-277768460>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABZrE-xKn0Vz7FJl95RCWeTj3dSfLWfEks5rZ2VcgaJpZM4L16j1>
> .
>

gandrewstone commented:
Thanks for finding and working on a solution to this issue!

Eliminating the ->show at line 548 means that the icon does not appear until the splash screen is complete.  I would like the icon to appear right away, since this is always a good indication that the program is actually running, if the splash screen gets hidden.

I can't reproduce on my machine (ubuntu MATE 16.04 -- I see the menu no problem), so I need to ask you to try 2 possibilities:
First, call show() twice (put line 548 back).  
If that doesn't work add a call to Hide() on line 560 (BEFORE "trayIcon->setContextMenu(trayIconMenu);").

Both of these options put the tray icon up quickly, but hopefully recalculate the tray object.  The second option is not as nice because the icon "blinks" once.  But you barely notice it even if looking so I still think that it is better than having no icon during the startup process.

jamoes commented:
Thank you for taking a look at this!

I am running Ubuntu 16.04 with the XFCE desktop environment. So, there must be something different about how XFCE and Unity handle context menus in tray icons compared to MATE.

I just tried both of your suggestions. The first suggestion (calling `show()` twice) unfortunately did not fix to issue for me. The second suggestion (calling `hide()` and then `show()` after the context menu is set) did fix the issue of the empty menu, but it also resulted in the icon blinking (disappearing and reappearing) during launch. I think the blinking is actually quite noticeable, because it results in other icons shifting around as well.

For comparison with other applications, I was able to find two applications on my system that have both a tray icon and a splash screen: LibreOffice (after enabling the "Enable systray Quickstarter" option), and qBittorrent. Both of these applications do not show the tray icon until after the splash screen is complete. Additionally, looking at Bitcoin Core's current behavior, the tray icon is not shown until after the splash screen is complete.

So, my preference would be to keep this change as-is: do not show the tray icon until the splash screen is complete. This seems consistent with other applications, and avoids the blinking tray icon during launch. Users can still see that the application is loading by the splash screen and the splash screen icon in the task bar.

gandrewstone commented:
ok last shot at making this perfect:  Try creating the menu during the tray icon creation, but keep the menu entry assignment where it is.  Google searching suggests that this will work...  Like this:

`void BitcoinGUI::createTrayIcon(const NetworkStyle *networkStyle)
{
#ifndef Q_OS_MAC
    trayIcon = new QSystemTrayIcon(this);
    QString toolTip = tr("Bitcoin Unlimited client") + " " + networkStyle->getTitleAddText();
    trayIcon->setToolTip(toolTip);
    trayIcon->setIcon(networkStyle->getTrayAndWindowIcon());

    trayIconMenu = new QMenu(this); // creation of the menu needs to be done when the icon is hidden
    trayIcon->setContextMenu(trayIconMenu);
    
    trayIcon->show();
#endif

    notificator = new Notificator(QApplication::applicationName(), trayIcon, this);
}

void BitcoinGUI::createTrayIconMenu()
{
#ifndef Q_OS_MAC
    // return if trayIcon is unset (only on non-Mac OSes)
    if (!trayIcon)
        return;

    if (!trayIconMenu) // only create if we didn't already create the menu when the tray icon was created.
      {
      trayIconMenu = new QMenu(this);
      trayIcon->hide();
      trayIcon->setContextMenu(trayIconMenu);
      trayIcon->show();
      }
.
.
.
`

jamoes committed:
[QT] Bugfix: ensure tray icon menu is not empty

jamoes commented:
That worked! Thank you! It's nice to have it perfect :)

I have applied the changes, and rebased with the `release` branch.

Note that I did not include the second `if (!trayIconMenu)` statement in the `createTrayIconMenu` method. This is because the `createTrayIcon` method is called from `BitcoinGUI`'s constructor, so it should always be set, and it seemed redundant to include the instantiation of the menu object twice. Instead, I just added a check for the existence of `trayIconMenu` to the sanity check, which returns if it is not set.

gandrewstone commented:
great!  WRT if(!trayIconMenu), thanks for analyzing the lack of a need for that.  I just quickly did something that was provably correct for the sake of testing...
